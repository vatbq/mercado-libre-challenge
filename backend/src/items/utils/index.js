const parseItem = (item) => {
  const [priceAmount, priceDecimal] = item.price.toString().split('.');

  return {
    id: item.id,
    title: item.title,
    price: {
      currency: item.currency_id,
      amount: priceAmount,
      decimals: priceDecimal,
    },
    picture: item.thumbnail,
    condition: item.condition,
    free_shipping: item.shipping.free_shipping,
  };
};

const parseDetailItem = (item) => ({
  ...parseItem(item),
  sold_quantity: item.initial_quantity,
  description: item.plain_text,
  categories: item.categories.path_from_root.map((category) => category.name),
});

module.exports = {
  parseDetailItem,
  parseItem,
};
