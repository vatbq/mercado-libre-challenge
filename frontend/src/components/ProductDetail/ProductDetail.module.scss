@import "styles/_global";
@import "styles/_colors";

.root {
  padding-left: $layout-padding-horizontal;
  padding-right: $layout-padding-horizontal;
  padding-bottom: 2rem;
}

.productContainer {
  display: flex;
  flex-direction: column;
  gap: 2rem;
  background-color: $white;
  padding: 2rem;
  border-radius: 0.25rem;
  margin-top: 1rem;
}

.productDetailContainer {
  display: flex;
}

.productImage {
  width: 680px;
  height: auto;
  min-width: 680px;
}

.productInfo {
  display: flex;
  flex-direction: column;
  gap: 2rem;
}

.price {
  font-size: 46px;
  margin: 0;
}

.buyButton {
  background-color: $button-color;
  color: $white;
  border-radius: 0.5rem;
  min-width: 4rem;
  padding: 1rem;
  border: none;
  font-size: 18px;
  width: 16rem;

  &:hover {
    background-color: darken($button-color, 10%);
    cursor: pointer;
  }
}

.title {
  margin: 0;
}

.descriptionContainer {
  max-width: 680px;
  display: flex;
  flex-direction: column;
  gap: 2rem;
}

.descriptionTitle {
  font-size: 28px;
  margin: 0;
}

.description {
  color: $text-color;
  margin: 0;
}

.condition {
  color: $text-color;
}
