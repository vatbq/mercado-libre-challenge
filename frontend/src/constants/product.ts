export const ProductCondition = {
  New: "new",
  Used: "used",
} as const
