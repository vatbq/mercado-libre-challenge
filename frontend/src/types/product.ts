export interface Product {
  id: string;
  title: string;
  price: {
    currency: string;
    amount: string;
    decimals: string;
  };
  picture: string;
  condition: string;
  free_shipping: boolean;
  sold_quantity: number;
  description: string;
  categories: string[];
}
