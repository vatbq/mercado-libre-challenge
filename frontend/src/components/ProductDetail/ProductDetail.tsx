import { Product } from "types/product";
import { getProductPrice } from "utils/product";
import Breadcrumb from "components/Breadcrumb";

import classes from "./ProductDetail.module.scss";

interface ProductProps {
  product: Product;
  categories: string[];
}

const ProductDetail = ({ product, categories }: ProductProps) => {
  const breadcrumbItems = categories.map((category) => ({ item: category }));
  const price = getProductPrice(product);

  return (
    <div className={classes.root}>
      {breadcrumbItems.length > 0 && <Breadcrumb items={breadcrumbItems} />}
      <div className={classes.productContainer}>
        <div className={classes.productDetailContainer}>
          <img
            src={product.picture}
            alt="Product cover"
            className={classes.productImage}
          />
          <div className={classes.productInfo}>
            <div>
              <p className={classes.condition}>
                {product.condition === "new" ? "Nuevo" : "Usado"} -{" "}
                {product.sold_quantity} vendidos
              </p>
              <h2 className={classes.title}>{product.title}</h2>
            </div>
            <p className={classes.price}>$ {price}</p>
            <button className={classes.buyButton}>Comprar</button>
          </div>
        </div>
        <div className={classes.descriptionContainer}>
          <p className={classes.descriptionTitle}>Descripci√≥n del producto</p>
          <p className={classes.description}>{product.description}</p>
        </div>
      </div>
    </div>
  );
};

export default ProductDetail;
