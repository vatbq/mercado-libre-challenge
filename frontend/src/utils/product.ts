import { Product } from "types/product";

const numberFormatter = new Intl.NumberFormat("en-US", {
  style: "decimal",
  minimumFractionDigits: 0,
  maximumFractionDigits: 2,
});

export const getProductPrice = ({ price }: Product): string => {
  const decimals = price.decimals ? parseFloat(price.decimals) : 0;
  return numberFormatter.format(parseFloat(`${price.amount}.${decimals}`));
};
