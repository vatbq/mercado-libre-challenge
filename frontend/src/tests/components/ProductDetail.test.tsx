import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import ProductDetail from "components/ProductDetail";
import { productMock } from "tests/factories/product";
import * as productUtils from "utils/product";

jest.mock("components/Breadcrumb", () => () => <div>Breadcrumb</div>);
jest.mock("utils/product", () => ({
  getProductPrice: jest.fn(),
}));

const categoriesMock = ["Electronics", "Gadgets"];

describe("<ProductDetail />", () => {
  beforeEach(() => {
    productUtils.getProductPrice.mockImplementation(() => "1,234.56");
  });

  it("renders ProgramCard unchanged", () => {
    const { container } = render(
      <ProductDetail product={productMock} categories={categoriesMock} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders the Breadcrumb component with the correct categories", () => {
    render(<ProductDetail product={productMock} categories={categoriesMock} />);

    expect(screen.getByText("Breadcrumb")).toBeInTheDocument();
  });

  it("displays the product title, condition, sold quantity, and price correctly", () => {
    render(<ProductDetail product={productMock} categories={categoriesMock} />);

    expect(screen.getByText(productMock.title)).toBeInTheDocument();
    expect(
      screen.getByText(
        `${productMock.condition === "new" ? "Nuevo" : "Usado"} - ${productMock.sold_quantity} vendidos`,
      ),
    ).toBeInTheDocument();
    expect(screen.getByText(`$ 1,234.56`)).toBeInTheDocument();
  });

  it("displays the product description", () => {
    render(<ProductDetail product={productMock} categories={categoriesMock} />);

    expect(screen.getByText("DescripciÃ³n del producto")).toBeInTheDocument();
    expect(screen.getByText(productMock.description)).toBeInTheDocument();
  });

  it("renders the product image with the correct src and alt attributes", () => {
    render(<ProductDetail product={productMock} categories={categoriesMock} />);

    const image = screen.getByRole("img");

    expect(image).toHaveAttribute("src", productMock.picture);
    expect(image).toHaveAttribute("alt", "Product cover");
  });
});
