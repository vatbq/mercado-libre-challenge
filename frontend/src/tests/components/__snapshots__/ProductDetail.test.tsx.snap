// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetail /> renders ProgramCard unchanged 1`] = `
<div>
  <div
    class="root"
  >
    <div>
      Breadcrumb
    </div>
    <div
      class="productContainer"
    >
      <div
        class="productDetailContainer"
      >
        <img
          alt="Product cover"
          class="productImage"
          src="https://example.com/product1.jpg"
        />
        <div
          class="productInfo"
        >
          <div>
            <p
              class="condition"
            >
              Nuevo
               -
               
              150
               vendidos
            </p>
            <h2
              class="title"
            >
              Wireless Bluetooth Headphones
            </h2>
          </div>
          <p
            class="price"
          >
            $ 
            1,234.56
          </p>
          <button
            class="buyButton"
          >
            Comprar
          </button>
        </div>
      </div>
      <div
        class="descriptionContainer"
      >
        <p
          class="descriptionTitle"
        >
          Descripci√≥n del producto
        </p>
        <p
          class="description"
        >
          High-quality wireless Bluetooth headphones with noise cancellation and up to 20 hours of battery life.
        </p>
      </div>
    </div>
  </div>
</div>
`;
